### ISSUE TITLE
Assigning parameter to tensor doesn't pass gradient info in Autograd

### ISSUE THREAD

**Issue by yonghakim** [pytorch/pytorch#103155]
### 🐛 Describe the bug

Hi,

seems this https://github.com/pytorch/pytorch/issues/77565 happens in autograd function. Can you please check?

```python
import torch


class Compress(torch.autograd.Function):

    @staticmethod
    def setup_context(ctx, inputs, output):
        # layer_info, datatype = inputs
        pass

    @staticmethod
    def forward(col_list):

        x_list = torch.ones((len(col_list), 1))
        for i in range(len(col_list)):
            x_list[i] = col_list[i]
        return x_list

# inside Autograd
col_list = [1, torch.tensor(2., requires_grad=True)]
x_list = Compress.apply(col_list)
print(x_list)

# outside
x_list = torch.ones((len(col_list), 1))
for i in range(len(col_list)):
    x_list[i] = col_list[i]

print(x_list)
```

> 
> tensor([[1.],
>         [2.]])
> tensor([[1.],
>         [2.]], grad_fn=\<CopySlices\>)
> 
> Process finished with exit code 0
> 

### Versions

Collecting environment information...
PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.102)
CMake version: version 3.24.4
Libc version: N/A

Python version: 3.10.10 | packaged by conda-forge | (main, Mar 24 2023, 20:12:31) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-12.5-arm64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] torch==2.0.0
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] torch                     2.0.0                    pypi_0    pypi

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by soulitzer** [comment]
Autograd custom Function does not introspect into lists to look for tensors that require grad, so in this case it is assuming that none of your inputs require grad. 

To work around this you can pass in your arguments as varargs instead:
```python
import torch


class Compress(torch.autograd.Function):

    @staticmethod
    def setup_context(ctx, inputs, output):
        # layer_info, datatype = inputs
        pass

    @staticmethod
    def forward(*col_list):

        x_list = torch.ones((len(col_list), 1))
        for i in range(len(col_list)):
            x_list[i] = col_list[i]
        return x_list

# inside Autograd
col_list = [1, torch.tensor(2., requires_grad=True)]
x_list = Compress.apply(*col_list)
print(x_list)

# outside
x_list = torch.ones((len(col_list), 1))
for i in range(len(col_list)):
    x_list[i] = col_list[i]

print(x_list)
```

**Comment by soulitzer** [comment]
Closing for now because it is expected behavior
